### PR TITLE
[undertow] Asynchronous request processing

### PR DESCRIPTION
**Pull request by carterkozak**
This is an implementation of the ListenableFuture based approach

Target generated interface:

```java
interface MyService {
  ListenableFuture<Foo> getFoo(AuthHeader authHeader);
}
```

impl:

```java
class MyResource {
  @Override
  ListenableFuture<Foo> getFoo(AuthHeader authHeader) {
    return Futures.immediateFuture(Foo.builder().build());
  }
}
```

Where the blocking implementation would take the form:

```java
interface MyService {
  Foo getFoo(AuthHeader authHeader);
}
```

Sample generated Endpoints:
https://github.com/palantir/conjure-java/blob/eb25f77a0bd28c04df894ffd7dce34498f19ce35/conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedEndpoints.java#L161-L181

==COMMIT_MSG==
The Undertow server target supports asynchronous request processing.
==COMMIT_MSG==

### CODE REVIEW

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java
Comment (bavardage) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: so on which thread does `BinaryResponseBody#write` end up being called?

I guess this is back on an io thread?

Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: `BinaryResponseBody#write` is always executed on the worker pool, blocking I/O is not allowed on the I/O threads, they're for non-blocking I/O only.

Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: A slow client will cause `BinaryResponseBody#write` to take a very long time, so it must run on the large pool.

Comment (bavardage) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: Ok - yeah - that makes sense. I now see where we dispatch back to the worker pool.

Comment (bavardage) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: just in https://github.com/palantir/conjure-java/pull/361/files#diff-c3b5cc874401c390b8d96224a8786de5R70 ?

Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: Yep, exactly

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (markelliot) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: is this what cancels work if the caller closes their connection?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureUndertowRuntime.java
Comment (markelliot) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureUndertowRuntime.java]: probably a bit more useful for the message to read `asyncTimeout is required`

**Comment by markelliot** [comment]
Looks cool! Should we consider having some type that we own internally, for purposes of discussion here, let's call it `Promise<T>` that we use instead of forcing use of Guava? I think it's possible we may want to limit how much we end up depending on 3rd party stuff in API implementations.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: Only after the timeout has elapsed. This prevents services which return a future that is never completed from leaving requests active indefinitely.

I can add a listener to cancel the future if the server detects that the client has closed a connection/stream, but that's an optimization we can add outside of the initial implementation.

**Comment by carterkozak** [comment]
I agree that we should avoid third party dependencies where it's feasible to do so, but the counterargument is that we may create N new standards to solve the same problem, and steepen the learning curve for developers to use this feature.

Any approach we take using the `Promise<T>` model (including ListenableFuture) requires developers to understand how to create and use a concrete `Promise` implementation. If we pick a type which developers are already familiar with, it will reduce friction. There's an additional cost to implement supporting functionality as well as teaching devs how to use it.

An alternative approach is to use java 8 [CompletableFuture](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html) which solves the same set of problems as ListenableFuture. The API isn't as elegant as the guava type, and integration with the fork-join pool may yield unexpected results, but we can add error-prone checks to disallow usage that results in execution on the fork-join pool. Unlike ListenableFuture, it's relatively difficult to propagate cancellation back through a chain of CompletableFutures, which will be problematic when we time out operations or detect dropped connections.

The benefit of Guavas ListenableFuture is the breadth of supporting functionality
https://github.com/google/guava/wiki/ListenableFutureExplained
Using guava functionality to create and transform futures results in free cancellation propagation.

**Comment by carterkozak** [comment]
@markelliot Most of my response was spent avoiding your question about implementing our own `Promise` type, and listed alternatives. The upside is greater control, and if we do it well, potentially a more ergonomic API (this worked well for `Refreshable<T>`). The place where I'm getting stuck is that this problem isn't sufficiently different from `ListenableFuture`, so I don't see a great deal of upside in building our own `Promise<T>` aside from shedding the guava dependency. The cost will be relatively high as we would need to reimplement guava functionality, as well as adapters between `Promise<T>` and guava types.

@bavardage @j-baker Do you have thoughts/opinions on the target return type?

**Comment by uschi2000** [comment]
I'm not actually terribly concerned with exposing ListenableFuture because these are not cross-service/repo APIs: the guava dependency would be localized to my service only, and consumers of my API don't have to know that internally I'm using a ListenableFuture.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: why false?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: there's a race here between isDone and cancel (unless you can convince me that no other thread can possibly access this attachment)

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: IMO, don't need the null checks

File: conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/UndertowRuntime.java
Comment (uschi2000) [inline conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/UndertowRuntime.java:40]: sync implementations don't need the async stuff. possible to separate?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java:123]: what happens if the future tries to write to the exchange after the exchange has been completed? presumably we'll get an error thrown on the future thread?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: why false?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java:136]: the javadoc indicates that you should check whether `isInIoThread`

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: no need to check whether the future is already done?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: I *think* I get it, but I'm not sure; can you try to reword this explanation?
(is the problem that we need to call exchange.dispatch before the root handler terminates so that the connection isn't closed?)

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: remove null check

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: remove null check

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: remove one +

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: I think we need to help users of the async framework define a sensible executor for their work. maybe a `ConjureAsyncRuntime` could provide such a thing?

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: I'd be interested in seeing the API that a service author would interact with. (I assume they won't get to see the exchange)

Comment (uschi2000) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: ignore me, saw the thing down below

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java
Comment (uschi2000) [inline conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: picking up the question from above: how am I expected to construct these futures? on what thread pool should I run them?

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: Thread interruption can result in unexpected behavior which I'd prefer to avoid forcing service implementors to handle. The common case for this feature is not for work running on a thread, but for queued async tasks where the interruption flag makes no difference.
I will extract a boolean constant and add a comment.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: The `isDone` check is unnecessary, `cancel(boolean)` is threadsafe and will no-op if the future has completed in some other way.

File: conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/UndertowRuntime.java
Comment (carterkozak) [reply conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/UndertowRuntime.java:40]: the `UndertowRuntime` interface is only meant to be consumed by generated code. Sync implementations will not be generated to invoke `async()`. I think additional separation will make the code unwieldy, but you may have something in mind that I haven't considered.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: same as the other future, replied above

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java:123]: Writing the error response back to a client cannot work if the request has completed, but that's no different from the blocking case when a client drops. Either way we log the failure gracefully.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: No, the `isDone` check in the completion listener was unnecessary, I have removed it. If the future is done, `cancel` is a no-op.

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java]: That is correct, I've updated the comment.  Does it make more sense this way?

```java
// Dispatch the registration task, this accomplishes two things:
// 1. Puts the exchange into a 'dispatched' state, otherwise the request will be terminated when
//    the endpoint HttpHandler returns. See Connectors.executeRootHandler for more information.
// 2. Avoids racing completion with the current handler chain. The dispatched task will be executed
//    after this Endpoints HttpHandler returns. Otherwise an unlucky future could race
//    HttpServerExchange.isInCall transitioning from true -> false, causing hte dispatch task not
//    to execute.
```

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java:136]: I will add class level javadoc describing the thread model. The description in HttpServerExchange.dispatch javadoc describes how to move work from an I/O thread to a task thread, but that's not what we want to accomplish here -- we already know we're on a task thread, but we need to mark the exchange as dispatched while we wait for the future to complete. I updated the comment, hopefully that makes the intent clearer.

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncTargetGeneratedService.java]: In the ideal case, you do not construct them yourself, but use a ListenableFuture from something else (a dialog async client future, [ListeningExecutorService](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/ListeningExecutorService.html)), perhaps with some transformation.

https://github.com/google/guava/wiki/ListenableFutureExplained#creation

The conjure async request processing feature is designed to allow any ListenableFuture to be sufficient regardless of how the service author constructs it. On completion, we dispatch back to the worker pool to serialize the result (success or failure) the same way we would in the blocking case, regardless of how and where the future executed. The only thing that we execute on the thread which completes the future is `exchange.dispatch(args)`, which sends work to the task pool.

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: Our server library already provides an executor factory, and service implementations don't get access to the `ConjureRuntime` and by extension the `ConjureAsyncRuntime`.

We shouldn't provide executors to services because we expect this feature to be used to reduce thread use. If the goal is to run work on another thread, then return the result, there's no reason to return a future, one can execute that work synchronously in the service method.

In some cases it may be useful for specific types of endpoints to use executors to allow certain types of tasks to queue for execution on a bounded number of threads, but I don't think this framework is the correct place to manage that executor.

**Comment by carterkozak** [comment]
ready for another review.

**Comment by carterkozak** [comment]
Feature complete implementation, still some work to clean things up a bit before it can be merged.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java
Comment (carterkozak) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java]: Looking for better ideas to opt into this feature.

My preference is to allow it to be configured per endpoint without requiring us to generate entire apis as either blocking or asynchronous. I don't want developers to split a single service API into multiple services for convenience when some endpoints benefit from asynchronous processing.

File: conjure-java-core/src/test/java/com/palantir/conjure/java/UndertowServiceEteTest.java
Comment (carterkozak) [inline conjure-java-core/src/test/java/com/palantir/conjure/java/UndertowServiceEteTest.java]: These belong in AsyncRequestProcessingTest, we don't need to modify this test.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java
Comment (uschi2000) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java]: I'm somewhat skeptical this is the right choice, because it complicates the conjure configuration space and the workaround is as simple as `return Futures.immediateFuture(syncReturnValue)`

**Review by uschi2000 — COMMENTED** [review]
fine with moving forward with this.

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: I think this is quite easy to abuse / needs a lot of understanding of undertow internals and threading strategies. Thoughts on how to avoid bad "pseudo-async" uses? Maybe we should write 2-3 examples of "good" and "bad".

Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: > I think this is quite easy to abuse / needs a lot of understanding of undertow internals and threading strategies

The intent of the design was to avoid needing to understand how the server works behind the scenes, and provide something that works as expected. The thread model I described should only need to be understood by developers writing code for the undertow generator and runtime.

Could you describe how this could be abused? I think there are ways it could be used inefficiently, but that's true of existing generator implementations as well.

I think I may have created confusion in this test, the handler implementation uses hand-written handler code because I had not written the generator yet at this point. All uses of the async component should be done through code we generate, which provides little to no room for abuse. I'm planning to replace these tests with a set which use generated handler code instead.

> Thoughts on how to avoid bad "pseudo-async" uses? Maybe we should write 2-3 examples of "good" and "bad".

Good idea, do you think the readme is a good place for these, or perhaps a separate `docs/undertow_async.md` document? Alternatively I can embed examples in javadoc.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java
Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java]: I agree that this is dirty, and I don't think it would be problematic for foundry if we used a global switch (e.g. `--undertowListenableFutures`).

I'm responsible for a project with ~1000 endpoints, the majority of which exist in a single module, ~60 of these endpoints require asynchronous request processing in order to avoid quickly saturating the server worker pool.

Thoughts on this approach?
1. Implement the `--undertowListenableFutures` CLI flag to make every method return ListenableFuture.
2. Implement second `--experimentalUndertowAsyncMarkers` flag which can be enabled to opt into asynchronous processing per-endpint. I will tighten the marker values to accept only an external-type-import annotation to avoid allowing markers to grow beyond existing usage on parameters.

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (uschi2000) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: I'd put it in the README. Naively, if I saw I should implement a
```
interface MyService {
  ListenableFuture<String> foo();
}
```
, I would make a `MyResource`, have it eat an ExecuterService, and then in my `MyServer` init method create something like a `Executors.newFixedThreadPool` with some arbitrary number, and then compute my futures on that thread pool. And then, I'd think, I'd have a perfectly async service with great performance. I imagine that's not what you want me to do :)

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java
Comment (uschi2000) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowTypeFunctions.java]: ok

File: conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java
Comment (carterkozak) [reply conjure-java-undertow-runtime/src/test/java/com/palantir/conjure/java/undertow/runtime/AsyncRequestProcessingTest.java]: Will do, thank you for the example :-)

**Comment by carterkozak** [comment]
Remaining work:
* ~~Update AsyncRequestProcessingTest to use generated code instead of adding cruft to UndertowServiceEteTest~~
* Provide readme examples of both good and bad uses of uses of this feature.

**Comment by carterkozak** [comment]
Ready for review.

**Comment by carterkozak** [comment]
Friendly ping for review @uschi2000 @markelliot @j-baker @bavardage

File: conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureUndertowRuntime.java
Comment (carterkozak) [inline conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureUndertowRuntime.java:82]: Do we want to increase this to match `ClientConfigurations.DEFAULT_READ_TIMEOUT` (5 minutes)? I chose 3 because it's a relatively low number, and it will be harder to decrease this moving forward than it is to increase. I also considered using 2 minutes based on https://github.com/palantir/conjure-java-runtime/pull/966

5 minutes is the safest bet to avoid breaking the proxy use case, but we may leverage this to reduce call times.

Thoughts?

**Review by uschi2000 — APPROVED** [review]

**Comment by carterkozak** [comment]
I am planning to merge this tomorrow morning (EDT) unless there is additional feedback. If you would like additional time please let me know.

File: readme.md
Comment (rubenfiszel) [inline readme.md:225]: Great docs!

**Review by rubenfiszel — APPROVED** [review]
As spoken offline, my personal preference would have had to use a generator flag to set the default mode, and endpoint markers to override. Currently, you can only override with "async" and there is no "sync" markers.
However, I understand that in order to disincentivize the use of markers, this implementation prefers the override to only go one way, from sync to async.

**Comment by rubenfiszel** [comment]
Really excited to see this feature in action!

**Pull request merged**
